src/pyproject_hooks/_in_process/_in_process.py

```diff
@@ -106,6 +106,18 @@ def find_spec(self, fullname, _path, _target=None):
 
         return spec
 
+    def find_distributions(self, context=None):
+        # Delayed import: Python 3.7 does not contain importlib.metadata
+        # If this method is being called it must be because
+        # `importlib.metadata`/`importlib_metadata` is available.
+        try:
+            from importlib_metadata import DistributionFinder, MetadataPathFinder
+        except ImportError:
+            from importlib.metadata import DistributionFinder, MetadataPathFinder
```

Comment:
Should we only define this if importlib.metadata is available?

I'm mildly concerned about importing `importlib_metadata` from here since pip would need to modify its vendored copy of pyproject-hooks to not import an installed module (we've had issues due to attempting to do this in the past, usually resulting in breakage of pip because the user somehow ended up with a bad copy of said package in their environment and the user needing to understand the nuances + details of how stuff works to fix their environment).

Reply:
I agree with @pradyunsg here. Pip would need to patch this out.

Can we not simply do a version check? Or even just prefer the stdlib version always? That way, pip (which only supports Python 3.8+) would never pick up `importlib_metadata`, avoiding any problems. Is there any justification for using `importlib_metadata` in preference to the stdlib implementation, when the latter is present?

Reply:
> Is there any justification for using importlib_metadata in preference to the stdlib implementation, when the latter is present?

IIRC, `importlib.metadata` was added to the stdlib, but then had significant API changes in Python 3.10, and `importlib_metadata` made those changes available to previous Python versions. So various tools using it prefer the backport over the stdlib module.

I also don't really like the idea that this will do different things depending on whether a separate package is installed or not. Maybe we should pretend that `importlib.metadata` was only added from Python 3.10 and do a version check?

(I remember this backporting causing issues before as well, though I don't recall which project that was in. I think for most stdlib modules doing backports like this would have been fine, but because `importlib.metadata` defines an interface *between* third party packages, the packages using it need to agree on whether they use the backport or not. With the benefit of hindsight, I think once it was in the stdlib, backports should have stopped, so everyone just used the stdlib version, even if that meant we had to wait longer to use new APIs.)

Reply:
Yeah, unfortunately `importlib_metadata` is not 100% backward compatible with `importlib.metadata`.

In general, I find that the following works fine:

* using `importlib_metadata.DistributionFinder, MetadataPathFinder` to implement `find_distrubitions`, but then using `importlib.metadata` high-level APIs

And the following does ***not*** work:

* using `importlib.metadata.DistributionFinder, MetadataPathFinder` to implement `find_distrubitions`, but then using `importlib_metadata` high-level APIs

(Of course that depends on the exact functionalities being used by all the other packages in the virtual environment)

That is the main reason why I preferred this order. 
But as discussed in https://github.com/pypa/pyproject-hooks/pull/195#discussion_r1586428386, I have no personal problem in exchanging the order in the `try...except...` if that is a better approach. Or even just using `importlib.metadata` all the time in Python 3.8+. Please let me know and I will proceed accordingly.

---

> Can we not simply do a version check? Or even just prefer the stdlib version always? That way, pip (which only supports Python 3.8+) would never pick up importlib_metadata, avoiding any problems.

I believe that preferring the `stdlib` version always (or not attempting to use `importlib_metadata` at all) will have the same limitations.

An anecdote about that:

Think about https://github.com/abravalheri/setuptools/pull/12 and the [patch necessary to make it work](https://github.com/abravalheri/setuptools/pull/12/files#diff-b64ef3dbcb447d2cfc0a81bcba734f9f7d58440adcc078fb23a31949b7c723d4R538-R539). What happens there is that the `except` branch is executed and `importlib.metadata` is always used.
However setuptools uses its own vendored version of `importlib_metadata` high-level API which causes the incompatibility documented in https://github.com/python/importlib_metadata/issues/486.

In fact any package using `importlib_metadata` is always subject to these problems as mentioned by Thomas in https://github.com/pypa/pyproject-hooks/pull/195#discussion_r1587193920.

Reply:
> Yeah, unfortunately importlib_metadata is not 100% backward compatible with importlib.metadata.

That's a problem for `importlib_metadata`/`importlib.metadata` IMO. But it's been discussed elsewhere and I don't want to reiterate old arguments here.

For this PR, the main point is that we do *not* want a situation where pip's behaviour can change as a result of the user installing `importlib_metadata` (whether it's a pristine copy, or some patched nightmare that a distro, in their wisdom, has created). That's something we try very hard to avoid in pip, because it has caused significant issues for us in the past - it's at the root of why [our policy on debundling](https://pip.pypa.io/en/stable/development/vendoring-policy/#debundling) is basically "please don't, but if you must, it's your problem to deal with".

Pip doesn't support Python 3.7, so I'm happy with any solution that only uses the stdlib `importlib.metadata` for Python 3.8+. At worst, it'll be pip patching this library when we vendor it, so nothing *has* to change here. But I'd expect other users of this library, like `build`, to have similar issues, and so I think it's worth discussing a solution that isn't pip-specific. But if the conclusion is "we don't want to bother about that", then that's OK - pip will patch it.

Reply:
I am happy to go with `importlib.metadata` always. That was more similar to my initial implementation.

> But I'd expect other users of this library, like build, to have similar issues, and so I think it's worth discussing a solution that isn't pip-specific.

I believe that `build` also uses the [same approach of preferring `importlib_metadata`](https://github.com/pypa/build/issues/692) and falling back to `importlib.metadata`. In this sense, the current implementation of this PR may be more compatible with `build`.

Reply:
I will open a separate PR with that approach so the maintainers can choose which one they prefer and close the other one.